packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/CopyFromRealmTests.kt

```diff
@@ -0,0 +1,755 @@
+        assertEquals("set-depth-2", managedObj.objectSetField.first().objectSetField.first().stringField)
+
+        val unmanagedCopy = managedObj.copyFromRealm(depth = 1u)
+        assertEquals("obj-depth-1", unmanagedCopy.nullableObject!!.stringField)
+        assertEquals("list-depth-1", unmanagedCopy.objectListField.first().stringField)
+        assertEquals("set-depth-1", unmanagedCopy.objectSetField.first().stringField)
+        assertNull(unmanagedCopy.nullableObject!!.nullableObject)
+        assertEquals(0, unmanagedCopy.objectListField.first().objectListField.size)
+        assertEquals(0, unmanagedCopy.objectSetField.first().objectSetField.size)
+    }
+
+    @Test
+    fun depth_primitiveListsAndSetsWhenDepthIsReached() {
+        val sample = Sample().apply {
+            stringField = "obj-depth-0"
+            stringListField = realmListOf("foo", "bar")
+            objectListField = realmListOf(
+                Sample().apply {
+                    stringField = "list-depth-1"
+                }
+            )
+            stringSetField = realmSetOf("foo", "bar")
+            objectSetField = realmSetOf(
+                Sample().apply {
+                    stringField = "set-depth-1"
+                }
+            )
+        }
+
+        val managedObj = realm.writeBlocking {
+            copyToRealm(sample)
+        }
+        val unmanagedCopy = managedObj.copyFromRealm(depth = 0u)
+        assertNull(unmanagedCopy.nullableObject)
+        assertEquals(0, unmanagedCopy.objectListField.size)
+        assertEquals(0, unmanagedCopy.objectSetField.size)
+        assertEquals(2, unmanagedCopy.stringListField.size)
+        assertEquals(2, unmanagedCopy.stringSetField.size)
+    }
+
+    @Test
+    fun linkingObjectsAreNotCopied() {
+        val sample = Sample().apply {
+            nullableObject = Sample()
+        }
+        val managedObj = realm.writeBlocking {
+            copyToRealm(sample)
+        }
+        assertEquals(1, managedObj.nullableObject!!.linkingObject.size)
+        val unmanagedCopy = managedObj.copyFromRealm()
+        assertFailsWith<IllegalStateException> {
+            unmanagedCopy.linkingObject // Empty RealmResults
+        }
+        assertFailsWith<IllegalStateException> {
+            unmanagedCopy.nullableObject!!.linkingObject // Have 1 backlink
+        }
+    }
+
+    @Test
+    fun managedObjectClosedAfterCopy() {
+        val sample = realm.writeBlocking {
+            copyToRealm(Sample().apply { stringField = "foo" })
+        }
+        sample.copyFromRealm(closeAfterCopy = true)
+        assertFalse(sample.isValid())
+        assertFailsWithMessage<IllegalStateException>("Cannot perform this operation on an invalid/deleted object") {
+            sample.stringField
+        }
+    }
+
+    // Verify that closing an object also release the underlying version
+    @Test
+    fun closedObjectsFreeVersion() = runBlocking {
```

Comment:
Playing a bit around with this I guess we are not really freeing the version of the source to `copyFromRealm` until they managed objects are GC'ed. They store a reference to the `RealmReference` and as this is tracked in the master `Realm` then it is not cleaned up before all objects are GC'ed too. Thus, this connection renders the `closeAfterCopy` useless, so I think we should remove that until we can guarantee immediate effect.